src/Mono.Android/Microsoft.Android.Runtime/ManagedValueManager.cs

```diff
@@ -20,10 +21,11 @@ class ManagedValueManager : JniRuntime.JniValueManager
 {
 	const DynamicallyAccessedMemberTypes Constructors = DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors;
 
-	Dictionary<int, List<IJavaPeerable>>?   RegisteredInstances = new Dictionary<int, List<IJavaPeerable>>();
+	Dictionary<int, List<GCHandle>>?   RegisteredInstances = new ();
 
-	internal ManagedValueManager ()
+	internal unsafe ManagedValueManager ()
 	{
+		JavaMarshal.Initialize (&FinishBridgeProcessing);
```

Comment:
@jonathanpeppers Note that the method that you register here will be called directly from the GC while the world is stopped. This means that you shouldn't register a managed method here. Instead there should be a native method that is not blocking (aka it dispatches the work of doing the java GC to a separate thread) and then returns so the C# GC can conclude. Once this separate thread finishes with the java gc, only then it can callback into managed so it does the `FinishCrossReferenceProcessing` (according to the new version of the api)

Reply:
I will revert https://github.com/dotnet/android/pull/10125/commits/f4a21484a015a5727a3593335b7841ab9f402ef5, I was already kind of skeptical that everything could be done in managed code.

/cc @grendello I think we need to be able to run these lines in our CoreCLR native code:

https://github.com/dotnet/android/blob/1f891246e2eb31756519494e6a8a087102de5681/src/native/mono/monodroid/osbridge.cc#L983-L989

Then at the end, it would call the UCO `FinishBridgeProcessing()` method.

Reply:
@BrzVlad how hard it is to register an `ecall` with CoreCLR? We can register a standard p/invoke method here, but `ecall` would be faster. I seem to recall it's not as straightforward in CoreCLR as it is in MonoVM, though?

Reply:
By `ecall` do you mean mono's `mono_add_internal_call` ? It is not clear to me how would you plan to use it in this context ?

Reply:
Yep, equivalent to Mono's `icall`. I assume there's a way to obtain its address from native code and use it to register in the above call.

Reply:
Maybe I'm missing some context, but `icalls` are used for managed code to call into native runtime code, that is registered to the runtime. While in `JavaMarshal.Initialize (..);` you are registering a native callback to be called by the runtime during GC. This native code is called directly, there should be no overhead or transitions involved, so I don't understand how an `icall` would be relevant here.

Reply:
OK, so it's me who misunderstood - I thought the callback would be called by the managed code at some point, not the native runtime. If it's a direct call from native runtime then the point is moot. Thanks for the clarification :) (I'm only now getting acquainted with this new GC bridge)